clang/lib/Basic/Targets/AArch64.cpp

```diff
@@ -202,19 +202,40 @@ AArch64TargetInfo::AArch64TargetInfo(const llvm::Triple &Triple,
 
 void AArch64TargetInfo::validatePAuthOptions(DiagnosticsEngine &Diags,
                                              const LangOptions &Opts) const {
-  if (Opts.SoftPointerAuth)
+  if (!Opts.PointerAuthCalls) {
+    if (Opts.PointerAuthInitFini)
+      Diags.Report(diag::warn_pauth_option_ignored) << "-fptrauth-init-fini";
+    if (Opts.FunctionPointerTypeDiscrimination)
+      Diags.Report(diag::warn_pauth_option_ignored)
+          << "-fptrauth-function-pointer-type-discrimination";
+    if (Opts.PointerAuthVTPtrAddressDiscrimination)
+      Diags.Report(diag::warn_pauth_option_ignored)
+          << "-fptrauth-vtable-pointer-address-discrimination";
+    if (Opts.PointerAuthVTPtrTypeDiscrimination)
+      Diags.Report(diag::warn_pauth_option_ignored)
+          << "-fptrauth-vtable-pointer-type-discrimination";
+    if (Opts.PointerAuthBlockDescriptorPointers)
+      Diags.Report(diag::warn_pauth_option_ignored)
+          << "-fptrauth-block-descriptor-pointers";
+  }
+
+  if (HasPAuth || Opts.SoftPointerAuth)
     return;
 
   bool NeedsFeatPAuth = false;
-  // FIXME: Some ptrauth_* intrinsics can be implemented without FEAT_PAuth.
+  // FIXME: Some ptrauth_* intrinsics can be implemented by only using
+  //        HINT-encoded instructions, thus not requiring FEAT_PAUTH.
   //        At now, checking conservatively.
   NeedsFeatPAuth |= Opts.PointerAuthIntrinsics;
   NeedsFeatPAuth |= Opts.PointerAuthCalls;
   NeedsFeatPAuth |= Opts.PointerAuthReturns;
   NeedsFeatPAuth |= Opts.getPointerAuthObjcIsaAuthentication() !=
                     PointerAuthenticationMode::None;
 
-  if (NeedsFeatPAuth && !HasPAuth)
+  // FIXME The particular command line option is not printed in error message,
```

Comment:
We already have a warning `warn_pauth_option_ignored` where we say that either `-fptrauth-calls` or `-mbranch-protection=pauthabi` should be enabled. Why not printing a similar error message mentioning both particular option (say, `-fptrauth-returns`) and the `-mbranch-protection=pauthabi` which includes the `-fptrauth-*` options? I don't feel that "misleadingly" in your comment fits here - one might want to enable only a particular feature and not the whole `-mbranch-protection` stuff, so, mentioning both in error message (like in `warn_pauth_option_ignored`) looks reasonable as for me

Reply:
Updated the message, thanks. Additionally, message duplication was fixed: now ignored/unsupported options are unset "on first iteration", so they do not trigger diagnostics anymore. Surprisingly, this slightly changed the behavior (due to `__has_feature(...)` preprocessor macro), but hopefully this is OK.